lib/std/ascii.zig

```diff
@@ -47,190 +50,170 @@ pub const control_code = struct {
 
     pub const DEL = 0x7F;
 
+    /// An alias to `DC1`.
     pub const XON = 0x11;
+    /// An alias to `DC3`.
     pub const XOFF = 0x13;
 };
 
-const tIndex = enum(u3) {
-    Alpha,
-    Hex,
-    Space,
-    Digit,
-    Lower,
-    Upper,
-    // Ctrl, < 0x20 || == DEL
-    // Print, = Graph || == ' '. NOT '\t' et cetera
-    Punct,
-    Graph,
-    //ASCII, | ~0b01111111
-    //isBlank, == ' ' || == '\x09'
-};
+// These naive functions are used to generate the lookup table
+// and they're used as fallbacks for if the lookup table isn't available.
+//
+// Note that some functions like for example `isDigit` don't use a table because it's slower.
+// Using a table is generally only useful if not all `true` values in the table would be in one row.
+
+fn isCntrlNaive(c: u8) bool {
+    return c <= control_code.US or c == control_code.DEL;
+}
+fn isAlphaNaive(c: u8) bool {
+    return isLower(c) or isUpper(c);
+}
+fn isXDigitNaive(c: u8) bool {
+    return isDigit(c) or
+        (c >= 'a' and c <= 'f') or
+        (c >= 'A' and c <= 'F');
+}
+fn isAlNumNaive(c: u8) bool {
+    return isDigit(c) or isAlphaNaive(c);
+}
+fn isPunctNaive(c: u8) bool {
+    @setEvalBranchQuota(3000);
+    return (c >= '!' and c <= '/') or
+        (c >= '[' and c <= '`') or
+        (c >= '{' and c <= '~') or
+        (c >= ':' and c <= '@');
+}
+fn isSpaceNaive(c: u8) bool {
+    @setEvalBranchQuota(5000);
+    return std.mem.indexOfScalar(u8, &spaces, c) != null;
+}
+
+/// A lookup table.
+const CombinedTable = struct {
+    table: [256]u8,
+
+    const Index = enum {
+        control,
+        alphabetic,
+        hexadecimal,
+        alphanumeric,
+        punct,
+        spaces,
+    };
 
-const combinedTable = init: {
-    comptime var table: [256]u8 = undefined;
+    /// Generates a table which is filled with the results of the given function for all characters.
+    fn getBoolTable(comptime condition: fn (u8) bool) [128]bool {
```

Comment:
This is pretty slick.  There is one potential improvement we could consider, namely, combine all the tables into 1 `u8` table.  If I remember correctly, an array of `bool` still takes 1-byte per element.   So we could represent up to 8 tables in the same amount of memory as 1 `bool` table.  The cost is likely a few more instructions (maybe just one) to access the data since we'd need to mask 1 specific bit.

Reply:
It sounds kind of useful but just so I understand the purpose correctly:

> There is one potential improvement we could consider, namely, combine all the tables into 1 `u8` table.

Right, that's what we're doing in `init`: we combine all tables into a single `[256]u8` table. Does your idea change the ultimate result of this single table that exists at runtime?
Will perhaps with your idea the runtime table instead be `[128]u8`?

>  If I remember correctly, an array of bool still takes 1-byte per element.

This seems to be correct.

> So we could represent up to 8 tables in the same amount of memory as 1 `bool` table.

So shouldn't that mean we would go from `[128]bool` to `[16]u8` because 128 / 8 = 16? Like we can represent the same data with `[128]bool` or `[16]u8`.

> The cost is likely a few more instructions (maybe just one) to access the data since we'd need to mask 1 specific bit.

Your idea overall sounds to me like an optimization that again reduces binary size at the cost of some speed. It would probably make sense to gate this behind `@import("builtin").mode == .ReleaseSmall` as well, like we're already doing it: we're omitting the table entirely from the binary in that build mode.

Reply:
If this is implemented, also consider `std.StaticBitSet` (from `lib/std/bit_set.zig`), which gives a named abstraction for exactly this kind of use case (I think).

Reply:
> Your idea overall sounds to me like an optimization that again reduces binary size at the cost of some speed. 

Here's some background on how I think this would affect memory/performance if you're interested.

Since there are 6 tables, if I'm understanding correctly, we'd go from 6 * 128 bytes down to 1 table of 128 bytes.  The extra cost might be a few extra instructions to access the data.  Note that on modern hardware, if our design saves us 1 cache miss, that means it would take hundreds OR thousands of extra instructions to be just as slow as that single cache miss (since cache misses are so slow).  Also note that most processors use 64-byte cache lines so this could be saving us 10 cache lines of memory, which is alot more potential for cache misses.

That's just some background on Data Oriented Design, the principles Andrew has been incorporating into the compiler recently and seeing big performance returns.  Checkout https://vimeo.com/649009599 if you haven't seen it.

In this case, my guess is saving that extra memory would not only lower the memory footprint, but also result in faster performance.  Of course performance depends on many things, especially the rest of the program and how much memory pressure there actually is.  But I've found that maintaining a smaller memory footprint tends to have alot more affect on performance than saving on instructions.  This is also just theory though, always benchmark to verify.

Reply:
> Since there are 6 tables, if I'm understanding correctly, we'd go from 6 * 128 bytes down to 1 table of 128 bytes.

But those 6 `[128]bool` tables only exist entirely at compile time. They never make it into the final runtime binary because all those 6 tables are temporary and are accumulated into a single `[256]u8` array. This sounds like a compile time optimization to me? I'm not sure if it makes sense to invest a lot of time into it because it doesn't affect runtime speed or how much memory is used at runtime.

Of course I would like to go from a `[256]u8` table to `[128]u8` because the remaining 128 bytes in that runtime table are always zero. These zeroes currently serve the purpose of returning `false` in all those cases where the character is outside of the 7-bit ASCII range. This is technically possible simply by going from `u8` to `u7` as I proposed here: https://github.com/ziglang/zig/pull/11629#issuecomment-1123305844
Sadly, as @ifreund explained, that is not (yet?) feasible.
But even so, 256 bytes is an okay price to say I think. I think it's a good balance between speed and memory usage. And as always in `ReleaseSmall`, the table doesn't even exist so I think that's compensated by that as well.

> That's just some background on Data Oriented Design, the principles Andrew has been incorporating into the compiler recently and seeing big performance returns. Checkout https://vimeo.com/649009599 if you haven't seen it.

That looks like an interesting talk. I might take a look. Thanks!

But apart from that, I think I get your point. It sounds good.
It's convenient that we have those bit sets in the `std` in case we need them.

Reply:
@marler8997: @r00ster91 is correct, the `[128]bool` arrays you are talking about won't be present in the compiled binary, they are just part of the code that goes into generating a single `[256]u8` which is already as compact as we can currently make it.

Reply:
> they are just part of the code that goes into generating a single [256]u8 which is already as compact as we can currently make it.

Ok great carry on then.  I stopped reading the code to soon.